src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -187,96 +187,99 @@ state switch
                     {
-                        // We're in either scenario 2 or 3. Do we have an existing tree to try replacing? Note: the file path here corresponds to Document.FilePath.
-                        // If a document's file path is null, we then substitute Document.Name, so we usually expect there to be a unique string regardless.
-                        var oldTree = compilationPair.CompilationWithoutGeneratedDocuments.SyntaxTrees.FirstOrDefault(t => t.FilePath == tree.FilePath);
-                        if (oldTree == null)
+                        // We do not have the exact document. It either means this document was recently added, or the
+                        // document was recently changed. We now need to update both the inProgressState and the
+                        // compilation. There are several possibilities we want to consider:
+                        //
+                        // 1. An earlier version of the document is present in the compilation, and we just need to update it to the current version
+                        // 2. The document wasn't present in the original snapshot at all, and we just need to add the
+                        //    document.
+                        // 3. The document wasn't present in the original snapshot, but an older file had been removed that
+                        //    had the same file path. As a heuristic, we remove the old one so we don't end up with
+                        //    duplicate documents.
+                        //
+                        // Note it's possible that we simply had never tried to produce a compilation yet for this project
+                        // at all, in that case GetPartialCompilationState would have produced an empty compilation, and it
+                        // would have updated inProgressProject to remove all the documents. Thus, that is no different than
+                        // the "add" case above.
+                        if (oldState != null)
                         {
-                            // Scenario 2.
-                            compilationPair = compilationPair.AddSyntaxTree(tree);
-                            inProgressProject = inProgressProject.AddDocuments([docState]);
+                            // Scenario 1. The document had been previously parsed and it's there, so we can update it with
+                            // our current state. Note if no compilation existed GetPartialCompilationState would have
+                            // produced an empty one, and removed any documents, so inProgressProject.DocumentStates would
+                            // have been empty originally.
+                            pendingActions = [(inProgressProject, new CompilationAndGeneratorDriverTranslationAction.TouchDocumentAction(oldState, docState))];
+                            inProgressProject = inProgressProject.UpdateDocument(docState, contentChanged: true);
```

Comment:
note: i think i'm passing in inProgressProject property to the  pending actions.  it represents the state prior to performing the action.

Reply:
and here we see teh core concept change.  instead of just manipulating the compilations *directly* (And thus needing the trees up front), we instead leverage totally normal pending actions, and we create an inprogress state that still has a *teeny* bit more work to do.  That work is going to be the same as today (namely parsing the syntax trees), but that work is now handled in the normal translation-step processing code we already have.

Also, *nicely*, instead of needing to get the syntax trees for the N linked-file siblings of this document *up front*, that work can be delayed until someone actually *needs* it (by asking for syntax trees for them, or for getting compilations for their projects).

Reply:
just a small thing? :)